### PR TITLE
Implements set_env_attr and get_env_attr for vector environments

### LINKED ISSUE #473: Set Attributes in Vector Environments
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [x] I have mentioned version numbers, operating system and environment, where applicable:

Operating System: Ubuntu Linux; Version: Commit c19876179a52acdc4d2e7dcb10dfbb4557adca86

Hey,

I am using tianshou as a submodule in a different project. There I use `ShmemVectorEnv` to simulate an environment. During training I would like to dynamically manipulate the environments by setting certain attributes to some values. Looking at the current code, this operation does not seem to be supported by the vector environments, is that correct? Is there some reason why `BaseVectorEnv` and the workers cannot implement a `__setattr__` method? 
Is there possibly some other solution to this problem or could such methods be implemented (I'd be happy to have a crack at it and raise a pull request)? Although I have not yet performed any benchmarks, I expect the cost of instantiating a new `ShmemVectorEnv` whenever I want to modify an attribute to be prohibitive.  
Thanks for your help!

Best,
Markus

### PR DESCRIPTION
**Pull request by Markus28**
- [x] I have marked all applicable categories:
    + [ ] exception-raising fix
    + [ ] algorithm implementation fix
    + [ ] documentation modification
    + [x] new feature
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make commit-checks` (**required**)
- [x] If applicable, I have mentioned the relevant/related issue(s)
- [x] If applicable, I have listed every items in this Pull Request below

This pull request aims to implement the features outlined in issue #473 

- Implement `get_env_attr`, `set_env_attr` for each worker class
- Implement `get_env_attr`, `set_env_attr` for vector environments
- Corresponding unit tests and documentation

### CODE REVIEW

**Comment by Markus28** [comment]
What's up with the failed profile check? I don't see where the segmentation fault could come from :(. Apart from that, how does it look?

**Comment by Trinkle23897** [comment]
not sure what's going wrong. I ran this test locally and it can successfully pass the check without segfault, maybe something went wrong in github container?

**Comment by Trinkle23897** [comment]
I found the cause: it is because torch==1.10.0

File: test/throughput/test_batch_profile.py [hunk 1]
```diff
@@ -106,7 +106,7 @@ def test_set_attr(data):
 
 def test_numpy_torch_convert(data):
     """Test conversion between numpy and torch."""
-    for _ in np.arange(1e5):
+    for _ in np.arange(1e4):  # not sure what's wrong in torch==1.10.0
```
Comment (Trinkle23897) [inline test/throughput/test_batch_profile.py:109]: here

File: tianshou/env/venvs.py [hunk 2]
```diff
@@ -130,17 +130,56 @@ def __getattribute__(self, key: str) -> Any:
...
+        self,
+        key: str,
+        id: Optional[Union[int, List[int], np.ndarray]] = None
+    ) -> List[Any]:
+        """Get an attribute from the underlying environments.
+
+        If id is an int, retrieve the attribute denoted by key from the environment
+        underlying the worker at index id. The result is returned as a list with one
+        element. Otherwise, retrieve the attribute for all workers at indices id and
+        return a list that is ordered correspondingly to id.
```
Comment (Trinkle23897) [inline tianshou/env/venvs.py:147]: should also mention if id is a list[int], same as set_env_attr

**Review by Trinkle23897 — APPROVED** [review]

**Comment by Markus28** [comment]
Awesome! Thanks for the quick review and the polishing :+1:

**Comment by ultmaster** [comment]
I have my own env worker here: https://github.com/microsoft/nni/blob/1eced0a7bff9d4de22487e7cfa45d41ce1f66605/nni/retiarii/strategy/_rl_impl.py#L25

But now it's broken because I need to implement two new methods.

I tried to update my code and delete the `__getattr__` method, but the updated version turns out to be imcompatible the tianshou 0.4.4.

How can I implement an env worker that works with both versions of tianshou? Implement both of `__getattr__` and `get_env_attr`? or use if-clause to check tianshou version?

----------------

My workaround here:

image

**Comment by Trinkle23897** [comment]
how about try...except ?

**Comment by ultmaster** [comment]
Not a good idea. Part of environment is user code. I don't want getattr to be run twice or catch unwanted errors.

**Comment by Trinkle23897** [comment]
Also a small request: can we have a unit test for your nni integration?

**Comment by ultmaster** [comment]
Sure. The related test code is here: https://github.com/microsoft/nni/blob/d50b46650bdc9e86aec8b2e0c07204c599310c9d/test/ut/retiarii/test_strategy.py#L144

I'm not sure whether it's handy right now because you need to install NNI from source (instead of from pip) for the tests.

-----

Or maybe tianshou can have a built-in support for Multi-thread env worker?

**Comment by ultmaster** [comment]
I'm experiencing issues as my CI upgrading to tianshou 0.4.6 breaks my env worker again. It looks like another interface change.

```
>   self.workers = [worker_fn(fn) for fn in env_fns]
E   TypeError: Can't instantiate abstract class MultiThreadEnvWorker with abstract method send
```

I understand that EnvWorker might not be considered as a public API. But as I'm implementing a 3rd-party env worker, it becomes a burden for me to be always compatible with all kinds of versions. I've already written 3 different version of env worker for tianshou 0.4.4 and 0.4.5 and 0.4.6, and unfortunately they are not compatible with each other.

**Comment by Trinkle23897** [comment]
Could you please upgrade to the newest version regardless of previous one? I'm super sorry about that happens but actually there's no CI that can secure this thing would never happen again.

**Comment by Trinkle23897** [comment]
> Or maybe tianshou can have a built-in support for Multi-thread env worker?

This is in [EnvPool](https://github.com/sail-sg/envpool/)

**Comment by ultmaster** [comment]
I've sent a patch. It's currently going through our CI.

microsoft/nni#4586

**Comment by ultmaster** [comment]
> > Or maybe tianshou can have a built-in support for Multi-thread env worker?
> 
> This is in [EnvPool](https://github.com/sail-sg/envpool/)

Looks like yet another framework under intense development. :)

**Comment by Trinkle23897** [comment]
I can make a post version tmr to fix this issue. (Also I just found another issue of consistency and it's critical to be fixed)

**Comment by Trinkle23897** [comment]
fixed in #536

**Pull request merged**